sdk/python/toolchain/pip.go

```diff
@@ -26,13 +28,125 @@ import (
 	"strings"
 
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
 )
 
 const (
 	windows             = "windows"
 	pythonShimCmdFormat = "pulumi-%s-shim.cmd"
 )
 
+type pip struct {
+	// The absolute path to the virtual env. Empty if not using a virtual env.
+	virtualenvPath string
+	// The virtual option as set in Pulumi.yaml.
+	virtualenvOption string
+	// The root directory of the project.
+	root string
+}
+
+var _ Toolchain = &pip{}
+
+func newPip(root, virtualenv string) (*pip, error) {
+	virtualenvPath := resolveVirtualEnvironmentPath(root, virtualenv)
+	logging.V(9).Infof("Python toolchain: using pip at %s", virtualenvPath)
+	return &pip{virtualenvPath, virtualenv, root}, nil
+}
+
+func (p *pip) InstallDependencies(ctx context.Context, cwd string, showOutput bool,
+	infoWriter io.Writer, errorWriter io.Writer,
+) error {
+	return InstallDependencies(
```

Comment:
The existing `InstallDependencies` can be inlined here, and there's more tidying up that can be done in this file, but the diff is already fairly large, so I am leaving that for a later PR.

Reply:
Makes sense, this PR is quite large to get through already :)